add freebsd version of wxStandardPaths::GetExecutablePath

sambler opened the pull request:
Hi, currently on FreeBSD `wxStandardPaths::GetExecutablePath` returns an empty string, this update fills in the right path.

Question - where would I add `-lprocstat` to `wx-config --libs` output when run on FreeBSD?

sambler committed:
add freebsd version of GetExecutablePath

vadz commented on src/unix/stdpaths.cpp:
```diff
@@ -167,6 +176,37 @@ wxString wxStandardPaths::GetExecutablePath() const
+        if (n_proc != 1)
+            break;
+
+        procstat_getpathname(ps, procs, pathname, sizeof(pathname));
```
Shouldn't we check the return value of this one, i.e. compare it with 0?

vadz commented on src/unix/stdpaths.cpp:
```diff
@@ -167,6 +176,37 @@ wxString wxStandardPaths::GetExecutablePath() const
+            break;
+
+        procstat_getpathname(ps, procs, pathname, sizeof(pathname));
+    } while (false);
```
The `do/while(false)` hack is rather ugly and unnecessary in C++ code. It would be nicer to rewrite this without, e.g.

```cpp
auto* const ps = procstat_open_sysctl();
if (!ps)
    return wxStandardPathsBase::GetExecutablePath();
wxON_BLOCK_EXIT1(procstat_close, ps);

auto* const procs = procstat_getprocs();
if (!procs)
    return wxStandardPathsBase::GetExecutablePath();
wxON_BLOCK_EXIT2(procstat_freeprocs, ps, procs);

if (procstat_getpathname(ps, procs, pathname, sizeof(pathname)) != 0)
    return wxStandardPathsBase::GetExecutablePath();

return pathname;
```

vadz reviewed (commented):
Thanks for adding this code, but, indeed, support for it needs to be added to the build system too:

1. You need to add new compile-time option `wxHAS_PROCSTAT` (you may choose a better name if you have one, of course), i.e. add it to the "Unix-specific options" section of `setup.h.in` and `build/cmake/setup.h.in`.
1. You should add `AC_CHECK_LIB(procstat, procstat_getpathname)` (this assumes that all the other functions are always available in that library if this one is) to `configure.ac` and add `-lprocstat` to `WXCONFIG_LIBS` and `AC_DEFINE(wxHAS_PROCSTAT)` if the check succeeds.
1. For CMake, I'm not sure how exactly to do it, to be honest, but I think you might use `FIND_LIBRARY()`.

sambler committed:
adjust configure.ac to test if functions available for procstat_* calls
add lib to wx-config --libs output
and define wxHAS_PROCSTAT in cxxflags

sambler committed:
use neater approach to steps taken
use wxON_BLOCK_EXIT to cleanup allocated structs
test for wxHAS_PROCSTAT being defined through configure steps

sambler committed:
add search path for system library

sambler committed:
first attempt at building with cmake
not complete yet

sambler commented:
I have refactored the `GetExecutablePath` code.
adjusted `configure.ac`, build using configure/make works for me.
started an attempt at adding into cmake but have not had any luck building with cmake yet. Any tips here would be appreciated.

vadz commented on src/unix/stdpaths.cpp:
```diff
@@ -31,12 +31,22 @@
     #include <unistd.h>
 #endif
 
+#if defined(__FreeBSD__)
```
I think this should be

```suggestion
#ifdef wxHAS_PROCSTAT
```

vadz commented on src/unix/stdpaths.cpp:
```diff
@@ -167,6 +177,28 @@ wxString wxStandardPaths::GetExecutablePath() const
 
     if ( !exeStr.empty() )
         return exeStr;
+#elif defined(__FreeBSD__) && wxHAS_PROCSTAT
```
And this

```suggestion
#elif defined(wxHAS_PROCSTAT)
```

vadz commented on src/unix/stdpaths.cpp:
```diff
@@ -167,6 +177,28 @@ wxString wxStandardPaths::GetExecutablePath() const
+
+    // procstat_freeprocs should be called before procstat_close
+    wxON_BLOCK_EXIT2(procstat_freeprocs, ps, procs);
+    wxON_BLOCK_EXIT1(procstat_close, ps);
```
This line needs to be moved above (before `procs` initialization), otherwise `procstat_close()` won't be called if `procstat_getprocs()` fails and we return early.

sambler replied:
`procstat_close` needs to be called after `procstat_freeprocs` so that `ps` is still valid in the `procstat_freeprocs` call. I was thinking code execution order, but the destructors would be run while clearing the stack, so that would be lifo making the first code line the last to run, which would work.

vadz replied:
Not sure if I'm missing something, but yes, exactly: because the dtors run in the opposite order, you have to put the guard for `procstat_close()` *before* the one for `procstat_freeprocs()`.

vadz commented on configure.ac:
```diff
@@ -7732,6 +7751,11 @@ case "$wxUSE_ZLIB" in
 esac
 
+if test "$wxUSE_PROCSTAT" = "yes"; then
+    WXCONFIG_CXXFLAGS="$WXCONFIG_CXXFLAGS -DwxHAS_PROCSTAT=1"
```
This shouldn't be necessary: `wxHAS_PROCSTAT` is only used internally.

```suggestion
```

vadz commented on build/cmake/setup.h.in:
```diff
@@ -621,6 +621,11 @@
+/*
+   On FreeBSD use libprocstat to get executable path name.
+*/
+#cmakedefine01 wxHAS_PROCSTAT
```
```suggestion
#cmakedefine wxHAS_PROCSTAT
```

vadz commented on configure.ac:
```diff
@@ -200,6 +200,7 @@ case "${host}" in
     AC_DEFINE(__FREEBSD__)
     AC_DEFINE(__BSD__)
     DEFAULT_DEFAULT_wxUSE_GTK=1
+    wxUSE_PROCSTAT=yes
```
I don't think this is needed at all, we already check for `host` before testing for procstat below.

vadz commented on configure.ac:
```diff
@@ -3424,6 +3425,24 @@ if test "$wxUSE_OPENGL" = "yes" -o "$wxUSE_OPENGL" = "auto"; then
+      *-*-freebsd*)
+        AC_MSG_CHECKING([for libprocstat])
+        AC_CHECK_LIB(procstat, procstat_getpathname, [
+                        PROCSTAT_LINK=" -L/usr/lib -lprocstat"
```
Is `-L/usr/lib` really needed? I'd expect the linked to look in this directory by default.

sambler replied:
When building static libs, the libs and app built and found the lib in `/usr/lib` but when building the shared libs linking failed with unable to find `libprocstat` so I added `-L/usr/lib`

vadz commented on build/cmake/build.cfg.in:
```diff
@@ -22,6 +22,7 @@ USE_GUI=@wxUSE_GUI_bf@
 USE_HTML=@wxUSE_HTML_bf@
 USE_MEDIA=@wxUSE_MEDIA_bf@
 USE_OPENGL=@wxUSE_OPENGL_bf@
+USE_PROCSTAT=@wxUSE_PROCSTAT_bf@
```
I don't think we need this.

vadz reviewed (commented):
Thanks, this looks much better, but there is some confusion between `wxUSE_PROCSTAT` and `wxHAS_PROCSTAT`: I've suggested the latter to test for procstat stuff presence in the library code, and this is the only place where it's needed, so there is no need to add it to `wx-config --cxxflags` output. Also, it's not a user-settable option (unless we want to allow disabling it by using `--disable-procstat` with configure, but I don't think it's worth it) it, so `wxUSE_PROCSTAT` is not needed at all. Finally, `wxHAS_XXX` constants are, by convention, either defined (as 1) or not, they should never be defined as 0.

As for CMake, I'm not sure what is the problem exactly?

P.S. Oh, sorry, and I forgot to say, but `configure` needs to be regenerated by running autoconf (ideally using `autoconf-for-wx` container with the fixed autoconf version, but you can do it using your autoconf version too, I'll just rerun it before merging then), otherwise your changes are not even tested in the FreeBSD CI job.

sambler commented:
I discovered the need to run autoconf, I didn't commit `configure` changes as there are a lot, I expect from different autoconf vers.

vadz commented:
> I discovered the need to run autoconf, I didn't commit `configure` changes as there are a lot, I expect from different autoconf vers.

Yes, this is why using the container image mentioned above is best, but if this can't be done, it's still better to regenerate it to at least test compiling this code in the FreeBSD CI job.

sambler committed:
remove use of wxUSE_PROCSTAT
remove define in cxxflags
rearrange placement of wxON_BLOCK_EXIT

sambler committed:
add generated configure

sambler commented:
Also of note, I'm not sure the lib tests need to go into configure, it would appear that freebsd is the only system using libprocstat, FreeBSD > 9.0 will always have it, searching man pages for procstat_* functions for netbsd, openbsd, dragonfly, darwin all come up empty.

sambler committed:
move freebsd check to base CMakeLists.txt

sambler committed:
cleanup unneeded cmake changes

MaartenBent commented on build/cmake/lib/base/CMakeLists.txt:
```diff
@@ -71,3 +71,8 @@ if(APPLE)
+
+if (${CMAKE_SYSTEM_NAME} STREQUAL "FreeBSD")
+    wx_lib_link_libraries(wxbase PUBLIC "-lprocstat")
+    target_compile_definitions(wxbase PRIVATE "-DwxHAS_PROCSTAT")
```
We normally don't use compile definitions. Everything is added to `setup.h[.in]`.
Most of the items there are named `HAVE_[NAME]`, so in this case ` HAVE_PROCSTAT` .

To integrate it in CMake, I think you can add a check in `setup.cmake` to check if the header is available:
```cmake
check_include_file(libprocstat.h HAVE_PROCSTAT)
```

This will set `HAVE_PROCSTAT` to 1 if it is available, and if `setup.h.in` contains this variable, it will be automatically set in the generated `setup.h`.

And then in `base/CmakeLists.txt`
```cmake
if (HAVE_PROCSTAT)
    wx_lib_link_libraries(wxbase PUBLIC "-lprocstat")
endif()
```

sambler replied:
Everything I can think of trying with cmake has failed, using cmake 3.31.6. I think mainly due to the inability to include `libprocstat.h` by itself. I tried `check_include_files` with the list of includes needed but that has failed as well. 

Using `find_library` in cmake does work, but still need to use `target_compile_definitions` to define it when compiling.

Every FreeBSD release since 9.0 (released 2012) has had `libprocstat` and headers. The searches I have done show FreeBSD as the only one using `libprocstat`. So just testing for `FreeBSD` will ensure the headers and lib exists, it isn't an optional component and no other system appears to have it.

MaartenBent replied:
You'll need to update `setup.h.in` too. With the patch below, the CMake build works for me (the library can be `PRIVATE` and doesn't need `-l` prefix). I also couldn't get `check_include_files` to work.

I used `wxHAVE_PROCSTAT` name because you also use this name in `stdpaths.cpp`.

For configure, I think you can just replace `CXXFLAGS="${CXXFLAGS} -DwxHAVE_PROCSTAT"` with `AC_DEFINE(wxHAVE_PROCSTAT)`, but I have not tested this.


```diff
 build/cmake/lib/base/CMakeLists.txt | 5 +++++
 build/cmake/setup.h.in              | 3 +++
 setup.h.in                          | 3 +++
 src/png                             | 2 +-
 src/webp                            | 0
 5 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/build/cmake/lib/base/CMakeLists.txt b/build/cmake/lib/base/CMakeLists.txt
index 7f60e93f527..32507f919c6 100644
--- a/build/cmake/lib/base/CMakeLists.txt
+++ b/build/cmake/lib/base/CMakeLists.txt
@@ -52,6 +52,11 @@ if(wxUSE_THREADS AND CMAKE_THREAD_LIBS_INIT)
     wx_lib_link_libraries(wxbase PRIVATE ${CMAKE_THREAD_LIBS_INIT})
 endif()
 
+if (${CMAKE_SYSTEM_NAME} STREQUAL "FreeBSD")
+    set(wxHAVE_PROCSTAT 1)
+    wx_lib_link_libraries(wxbase PRIVATE "procstat")
+endif()
+
 if(APPLE)
     wx_lib_link_libraries(wxbase
         PUBLIC
diff --git a/build/cmake/setup.h.in b/build/cmake/setup.h.in
index 952dcb68a49..d469368809b 100644
--- a/build/cmake/setup.h.in
+++ b/build/cmake/setup.h.in
@@ -908,6 +908,9 @@
 /* Define if you have wcsnlen() function */
 #cmakedefine wxHAVE_WCSNLEN 1
 
+/* define if you have procstat (FreeBSD) */
+#cmakedefine wxHAVE_PROCSTAT 1
+
 /* The number of bytes in a wchar_t.  */
 @SIZEOF_WCHAR_T_CODE@
 
diff --git a/setup.h.in b/setup.h.in
index 5d65c131a09..870cf9819eb 100644
--- a/setup.h.in
+++ b/setup.h.in
@@ -907,6 +907,9 @@
 /* Define if you have wcsnlen() function */
 #undef wxHAVE_WCSNLEN
 
+/* define if you have procstat (FreeBSD) */
+#undef wxHAVE_PROCSTAT
+
 /* The number of bytes in a wchar_t.  */
 #undef SIZEOF_WCHAR_T

```

sambler committed:
adjust define name missed earlier

sambler committed:
rename use procstat to have procstat

sambler committed:
add generated configure

sambler committed:
use AC_DEFINE in configure.ac
move FreeBSD test into setup.cmake to get the define in setup.h

sambler committed:
keep link libraries in the base CMakeLists.txt

sambler commented:
@MaartenBent using set have procstat works, only if it is in `setup.cmake`.

I see two issues, the second would need to be addressed in a different issue.

1. `wx-config` output. When I build static libs `-lprocstat` is in the `--libs` output, when I build shared libs it is not. This applies to both cmake and autoconf builds.

This means with static libs, building the samples, demos and utils works. (except wxrc which I need to manually add `-lprocstat` to `LDFLAGS` to build)

With shared libs, the libs build but the apps fail without `-lprocstat` in `LDFLAGS`.

2. I can't get cmake to compile the shared libs. The static libs do build (not in one pass).

This would appear to be dependencies, shared libs fail to link with libs that haven't compiled yet. Although, base linking to gtk libs doesn't sound right to me.

```
[  4%] Linking CXX shared library ../../lib/libwx_baseu-3.3.so
ld: error: unable to find library -lwx_gtk3u_xrc-3.3
ld: error: unable to find library -lwx_gtk3u_html-3.3
ld: error: unable to find library -lwx_gtk3u_qa-3.3
ld: error: unable to find library -lwx_gtk3u_core-3.3
ld: error: unable to find library -lwx_baseu_xml-3.3
ld: error: unable to find library -lwx_baseu_net-3.3
ld: error: unable to find library -lwx_baseu-3.3
```

MaartenBent commented:
>  using set have procstat works, only if it is in setup.cmake

I guess this makes sense. `lib/base/CMakeLists.txt` is in a sub_directory, so the define might be in the wrong scope. Not sure why it worked for me, maybe an old cached value. Or maybe it didn't work at all and compiled without procstat. I'll try to test it again some time.

I have no idea what is happening with the libraries it tries to link to. For me building the `wxbase` and `console` sample targets worked fine and had no link issues.

I did have to add some extra CMake flags, similar to what is needed in CI, to make everything work, but this was related to finding system libraries, not wx libraries:
https://github.com/wxWidgets/wxWidgets/blob/5bc736df13144800b64b034399d30a6111ee9cea/.cirrus.yml#L67-L71
